starcounter-include.html

```diff
@@ -238,22 +248,43 @@
          */
         StarcounterIncludePrototype._compositionChanged = function(compositionString) {
             if (compositionString) {
-                this._forceLayoutChange = true;
-                this._renderCompositionChange("explicit or fallback", () => this.stringToDocumentFragment(compositionString));
+                this.explicitComposition = compositionString;
+                return this._renderCompositionChange(compositionString, () => this.stringToDocumentFragment(compositionString));
             }
             else {
                 const compositionProvider = this.partial && findCompositionProvider(this.partial);
 
                 if (compositionProvider) {
                     this.partialId = compositionProvider.PartialId;
+                    this.explicitComposition = null;
                     if (compositionString === "") {
                         //this is a request from starcounter-layout-html-editor to reset to default composition
-                        this.Composition$ = "";
                     }
                     else if (compositionProvider.Composition$) {
                         return this._renderCompositionChange(compositionProvider.Composition$, () => {
                             this.storedLayout = compositionProvider.Composition$; //should always be string
-                            return this.stringToDocumentFragment(this.storedLayout);
+                            const merged = this.stringToDocumentFragment(this.storedLayout);
+
+                            if (compositionProvider.ViewUris) {
+                                const keys = Object.keys(this.partial);
+                                for(let key of keys) {
+                                    const scoped = this.partial[key];
+                                    if (scoped && scoped.Html && compositionProvider.ViewUris.indexOf(scoped.Html) === -1) {
+                                        if (this.defaultComposition) {
+                                            this.template.scopedNodes.forEach((nodes) => {
```

Comment:
Why are you traversing all the nodes? Wouldn't it be easier to traverse `defaultComposition` it already contains only filtered declarative shadow roots. We could change it to or add a map with a `scope` as a key.

Reply:
Yes, perhaps we can do that. Do you want to improve it?